### ISSUE TITLE
ReflectiveTypeAdapterFactory.BoundField.write memory overhead

### ISSUE THREAD

**Issue by sembseth** [google/gson#2321]
# Problem solved by the feature
Reduced memory allocation when serializing data.

# Feature description
Whenever a BoundField is written a new TypeAdapterRuntimeTypeWrapper is constructed. As far as I can tell the TypeAdapterRuntimeTypeWrapper could just as well be constructed as a final variable outside of the write method. When re-using Gson instances this will reduce the total amount of memory allocated during serialization.

![gson_memory_overhead](https://user-images.githubusercontent.com/125864935/220053447-31ba3fe4-2cd5-4fab-9f04-3a70543bc83b.png)

**Comment by Marcono1234** [comment]
Thanks! This looks like a good suggestion, do you want to create a pull request for this?
I assume if `serialized == false` there would also be no need to create the `TypeAdapterRuntimeTypeWrapper` then if it is not used anyways.

**Comment by eamonnmcmanus** [comment]
I wasn't completely convinced at first: if we construct the `TypeAdapterRuntimeTypeWrapper` outside the anonymous `BoundField` then we save having to construct it inside, but on the other hand we have a bit more memory being used all the time versus just inside `BoundField.write`. But actually I see that there are _three_ local variables captured in `createBoundField` just so that they can be used in this one line: `typeAdapter`, `context`, and `fieldType`. That means three hidden fields in the anonymous subclass. If we construct the final `TypeAdapter` just once then that is just one hidden field, assuming we also use it in the `readIntoArray` and `readIntoField` methods. So I think the extra memory is pretty small.

In short, thank you for the idea, and I agree with @Marcono1234 that a PR would be welcome.
